### PR TITLE
Consolidate strict code

### PR DESCRIPTION
**Pull request by kaidokert**
## Summary by Sourcery

Consolidate and optimize Montgomery arithmetic implementations across strict, constrained, and basic variants, adding a reusable strict Montgomery multiplication helper and improving reduction performance and documentation.

New Features:
- Introduce a public strict_montgomery_mul helper to multiply values already in Montgomery form without recomputing parameters.

Enhancements:
- Optimize Montgomery reduction in strict, constrained, and basic variants by using bitmasking and shifts instead of generic modulo operations, including a fast path for r_bits == 0.
- Refine strict and constrained modular exponentiation to reuse the new Montgomery multiplication helper, avoid redundant parameter computation, and adjust trait bounds accordingly.
- Clarify and expand documentation comments around Hensel lifting, N' computation methods, and Montgomery exponentiation behavior.

Tests:
- Replace the internal strict Montgomery multiplication test with a comprehensive suite of tests for the new strict_montgomery_mul, covering basic cases, edge conditions, overflow scenarios, multiple moduli, and consistency checks.

## Summary by CodeRabbit

- **New Features**
  - Added a public Montgomery multiplication helper and an exponentiation variant that lets you choose the N' computation method.

- **Performance**
  - Reduction now uses mask-and-shift (bitwise) operations and preserves Montgomery form through exponentiation for better efficiency.

- **API Changes**
  - Adjusted exponentiation signature and broadened generic constraints to support bitwise operations.

- **Bug Fixes**
  - Canonicalized parameters into [0,R) and added fast-paths/guards for r_bits == 0.

- **Refactor**
  - Consolidated multiplication paths behind the new helper.

- **Chores**
  - Removed an internal helper/tests and added extensive Montgomery tests.

### CODE REVIEW

File: modmath/src/montgomery/strict_mont.rs [hunk 1]
```diff
@@ -244,6 +246,32 @@ where
     Some((r, r_inv, n_prime, r_bits))
 }
 
+/// Montgomery multiplication (Strict): (a * R) * (b * R) -> (a * b * R) mod N
+/// Multiplies two values already in Montgomery form and returns result in Montgomery form
+pub fn strict_montgomery_mul<T>(a_mont: T, b_mont: &T, modulus: &T, n_prime: &T, r_bits: usize) -> T
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/montgomery/strict_mont.rs]: handled

File: modmath/src/montgomery/strict_mont.rs [hunk 2]
```diff
@@ -494,73 +525,12 @@ where
...
-    // Montgomery multiplication in Montgomery domain
-    // a_mont and b_mont are already in Montgomery form
-    // Result should be (a * b * R) mod N (still in Montgomery form)
-
-    // Step 1: Regular modular multiplication: (a_mont * b_mont) mod N
-    let product = crate::mul::strict_mod_mul(a_mont, b_mont, modulus);
-
-    // Step 2: Apply Montgomery reduction once to get (a * b * R) mod N
-    strict_from_montgomery(product, modulus, n_prime, r_bits)
+    strict_montgomery_mod_exp_with_method(
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/montgomery/strict_mont.rs]: this is handled in the new commit i think

File: modmath/src/montgomery/strict_mont.rs [hunk 3]
```diff
@@ -261,27 +289,29 @@ where
...
-    // 1. R = 2^r_bits
-    // 2. m = (a_mont * N') mod R
-    // 3. t = (a_mont + m * N) / R
+    // 1. R = 2^r_bits, mask = R - 1 = (1 << r_bits) - 1
+    // 2. m = ((a_mont & mask) * N') & mask  [only low bits, no expensive modulo!]
+    // 3. t = (a_mont + m * N) >> r_bits     [bit shift, no division!]
     // 4. if t >= N then return t - N else return t
 
-    let r = T::one() << r_bits; // R = 2^r_bits
+    let mask = (T::one() << r_bits) - T::one(); // mask = 2^r_bits - 1
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/montgomery/strict_mont.rs]: handled

**Comment by kaidokert** [comment]
@sourcery-ai review

File: modmath/src/montgomery/basic_mont.rs [hunk 4]
```diff
@@ -240,24 +246,33 @@ where
...
-    // 1. R = 2^r_bits
-    // 2. m = (a_mont * N') mod R
-    // 3. t = (a_mont + m * N) / R
+    // 1. mask = 2^r_bits - 1
+    // 2. m = ((a_mont & mask) * N') & mask  [only low bits, no expensive modulo!]
+    // 3. t = (a_mont + m * N) >> r_bits     [bit shift, no division!]
     // 4. if t >= N then return t - N else return t
 
-    let r = T::one() << r_bits; // R = 2^r_bits
+    // Check for r_bits == 0 to avoid underflow in mask calculation
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/montgomery/basic_mont.rs]: done

**Comment by kaidokert** [comment]
@sourcery-ai review

File: modmath/src/montgomery/strict_mont.rs [hunk 5]
```diff
@@ -244,6 +252,32 @@ where
...
+        + core::ops::Sub<Output = T>
+        + core::ops::Shr<usize, Output = T>
+        + core::ops::Shl<usize, Output = T>
+        + num_traits::ops::overflowing::OverflowingAdd
+        + num_traits::ops::overflowing::OverflowingSub,
+    for<'a> T: core::ops::Mul<&'a T, Output = T>
+        + core::ops::RemAssign<&'a T>
+        + core::ops::Sub<&'a T, Output = T>,
+    for<'a> &'a T: core::ops::Rem<&'a T, Output = T>
+        + core::ops::Sub<&'a T, Output = T>
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/montgomery/strict_mont.rs]: @gemini-code-assist No it's needed, because we are calling mod_mul - transitive dependency

File: modmath/src/montgomery/constrained_mont.rs [hunk 6]
```diff
@@ -227,21 +228,36 @@ where
...
-    let temp_sum = a_mont.wrapping_add(&temp_prod);
-    let t = temp_sum >> r_bits; // Divide by R = 2^r_bits
+    // Step 1: m = ((a_mont & mask) * N') & mask
+    let a_low = &a_mont & &mask;
+    let product = a_low * n_prime;
+    let m = &product & &mask;
+
+    // Step 2: t = (a_mont + m * N) >> r_bits
+    let m_times_n = m * modulus;
+    let temp_sum = a_mont.wrapping_add(&m_times_n);
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply modmath/src/montgomery/constrained_mont.rs]: @gemini-code-assist wrong as with your other comment on  sub, it takes a reference
